// @angulartoolsdr/shared-utils/esm2022/lib/pipes/cpf-cnpj-format.pipe.mjs
import { Pipe } from '@angular/core';
import IMask from 'imask';
import { Mask } from '../mask/mask';
import * as i0 from "@angular/core";
export class CpfCnpjFormat {
    constructor() {
        this.maskCnpj = IMask.createMask(Mask.getMaskCnpj());
        this.maskCpf = IMask.createMask(Mask.getMaskCpf());
    }
    transform(value) {
        if (value.length === 11) {
            this.maskCpf.resolve(value);
            return this.maskCpf.value;
        }
        else if (value.length === 14) {
            this.maskCnpj.resolve(value);
            return this.maskCnpj.value;
        }
        return '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: CpfCnpjFormat, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.9", ngImport: i0, type: CpfCnpjFormat, isStandalone: true, name: "cpfCnpjFormat" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: CpfCnpjFormat, decorators: [{
            type: Pipe,
            args: [{
                    name: 'cpfCnpjFormat',
                    standalone: true
                }]
        }] });
