// @angulartoolsdr/shared-utils/esm2022/lib/pipes/date-format.pipe.mjs
import { DatePipe } from '@angular/common';
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@ngx-translate/core";
export class DateFormatPipe {
    constructor(translateService) {
        this.translateService = translateService;
    }
    transform(value, pattern = 'mediumDate') {
        const datePipe = new DatePipe(this.translateService.currentLang);
        return datePipe.transform(value, pattern);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: DateFormatPipe, deps: [{ token: i1.TranslateService }], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.9", ngImport: i0, type: DateFormatPipe, isStandalone: true, name: "dateFormat" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: DateFormatPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'dateFormat',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i1.TranslateService }] });
