// @angulartoolsdr/shared-utils/esm2022/lib/pipes/utcDate.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class UtcDatePipe {
    transform(value) {
        if (!value) {
            return '';
        }
        const dateValue = new Date(value);
        const dateWithNoTimezone = new Date(dateValue.getUTCFullYear(), dateValue.getUTCMonth(), dateValue.getUTCDate(), dateValue.getUTCHours(), dateValue.getUTCMinutes(), dateValue.getUTCSeconds());
        return dateWithNoTimezone;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: UtcDatePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.9", ngImport: i0, type: UtcDatePipe, isStandalone: true, name: "utcDate" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: UtcDatePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'utcDate',
                    standalone: true
                }]
        }] });
