// @angulartoolsdr/shared-utils/esm2022/lib/services/alert.service.mjs
import Swal from 'sweetalert2';
import { inject, Injectable } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import * as i0 from "@angular/core";
export class AlertService {
    constructor() {
        this.translate = inject(TranslateService);
    }
    alertText(title, message, fnConfirm, fnCancel) {
        Swal.fire({
            title: title,
            text: message,
            icon: 'success',
            confirmButtonColor: '#0CC27E',
            confirmButtonText: '<i class="ft-check-circle"></i> ' + this.translate.instant('OK')
        }).then((result) => {
            if (result.value) {
                fnConfirm();
            }
            else if (fnCancel !== undefined) {
                fnCancel();
            }
        });
    }
    confirmText(title, message, fnConfirm, fnCancel) {
        let mensagem = '';
        let params = {};
        if (message instanceof Array) {
            mensagem = message[0];
            params = message[1];
        }
        else {
            mensagem = message;
        }
        Swal.fire({
            title: title,
            text: mensagem,
            icon: 'question',
            showCancelButton: true,
            confirmButtonColor: '#0CC27E',
            cancelButtonColor: '#FF586B',
            confirmButtonText: '<i class="ft-check-circle"></i> ' + this.translate.instant('SIM'),
            cancelButtonText: '<i class="ft-x-circle"></i> ' + this.translate.instant('NAO'),
        }).then((result) => {
            if (result.value) {
                fnConfirm();
            }
            else if (fnCancel !== undefined) {
                fnCancel();
            }
        });
    }
    confirmTextResetPassword(fnConfirm, fnCancel) {
        const swal = {
            title: this.translate.instant('RESETAR_SENHA'),
            text: this.translate.instant('DIGITE_CODIGO_AUTENTICACAO'),
            icon: 'info',
            input: 'text',
            inputValidator: (value) => {
                return !value && this.translate.instant('CAMPO_OBRIGATORIO');
            },
            showCancelButton: true,
            confirmButtonColor: '#0CC27E',
            cancelButtonColor: '#FF586B',
            cancelButtonText: '<i class="ft-x"></i> ' + this.translate.instant('CANCELAR'),
            confirmButtonText: '<i class="ft-check"></i> ' + this.translate.instant('OK'),
        };
        Swal.fire(swal).then((result) => {
            if (result.value) {
                fnConfirm(result.value);
            }
            else if (fnCancel !== undefined) {
                fnCancel();
            }
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: AlertService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: AlertService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: AlertService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
