// @angulartoolsdr/shared-utils/esm2022/lib/directives/slimscroll.directive.mjs
import { Directive } from '@angular/core';
import Scrollbar from 'smooth-scrollbar';
import * as i0 from "@angular/core";
export class SlimscrollDirective {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
    }
    ngAfterViewInit() {
        // smooth scroll
        Scrollbar.init(this.el.nativeElement);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: SlimscrollDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.9", type: SlimscrollDirective, isStandalone: true, selector: "[appSlimScroll]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.9", ngImport: i0, type: SlimscrollDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[appSlimScroll]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.Renderer2 }] });
